### PR TITLE
⚠️  Allow policy by default

### LINKED ISSUE #1726: When a dataset doesn't have PII columns, Fybrik policies prevent access to data
When no columns is marked as PII sensitive, the Fybrik roles samples prevent access to teh data set.

Here is our read example:
```
rule[{"action": {"name":"RedactAction", "columns": column_names}, "policy": description}] {
  description := "Redact columns tagged as PII in datasets tagged with finance = true"
  input.action.actionType == "read"
  input.resource.metadata.tags.finance
  column_names := [input.resource.metadata.columns[i].name | input.resource.metadata.columns[i].tags.PII]
  count(column_names) > 0
}
```
in the described case, `column_names` is null so there is no RedactAction and Fybrik behaves according to the default rule which means "Deny"

### PR DESCRIPTION
**Pull request by tomersolomon1**
Change the default policy to be allowed.

Fixes #1726

Signed-off-by: Tomer Solomon <tomer.solomon1@gmail.com>

### CODE REVIEW

File: site/docs/tasks/using-opa.md [hunk 1]
```diff
@@ -44,12 +44,12 @@ An example of an object that may be returned by the rule above:
 
 ## Fybrik Default Policies
 
-Fybrik ***denys by default*** any request if no rule is triggered. This behavior can be changed to ***allow by default*** by creating the following rule and upload it to OPA using methods described in this page:
+Fybrik ***allows by default*** any request if no rule is triggered. This behavior can be changed to ***deny by default*** by creating the following rule and upload it to OPA using methods described in this page:
```
Comment (shlomitk1) [inline site/docs/tasks/using-opa.md:47]: Creating a new rule does not change to deny. You need to replace the existing `rule [{}] { true }` with  `rule [{}] { false }` or removing it completely.

File: site/docs/tasks/using-opa.md [hunk 2]
```diff
@@ -44,12 +44,12 @@ An example of an object that may be returned by the rule above:
...
 
 ```yaml
 package dataapi.authz
 
-rule [{}] { true }
+rule [{}] { false }
 ```
 
 The verdict `allow` will be reached only if the conditions hold, and no other rule has been triggered, e.g. a rule requiring column redaction.
```
Comment (shlomitk1) [inline site/docs/tasks/using-opa.md:55]: The verdict `allow` will be reached if no other rule has been triggered, e.g. a rule requiring column redaction, or `deny`.

File: charts/fybrik/files/opa-server/policy-lib/default_policy.rego [hunk 3]
```diff
@@ -15,4 +15,4 @@ verdict[outputFormatted] {
 	outputFormatted := object.union({"action": actionFormatted}, outputWithoutAction)
 }
 
-rule[{}] { false }
+rule[{}] { true }
```
Comment (shlomitk1) [inline charts/fybrik/files/opa-server/policy-lib/default_policy.rego:18]: Remove 
```
verdict[output] {
	count(rule) == 0
	output = {"action": {"name":"Deny", "Deny": {}}, "policy": "Deny by default"}
}
```

File: site/docs/tasks/using-opa.md (hunk 2 above)
Comment (tomersolomon1) [reply site/docs/tasks/using-opa.md:55]: See now.

File: charts/fybrik/values.yaml [hunk 4]
```diff
@@ -442,6 +442,13 @@ opaServer:
   # Bootstrap policies to load upon startup
   bootstrapPolicies: {}
 
+  # Controls if to allow by default or not.
```
Comment (shlomitk1) [inline charts/fybrik/values.yaml:445]: # Controls the default behavior when no policies are defined, or none apply.

File: charts/fybrik/values.yaml [hunk 5]
```diff
@@ -442,6 +442,13 @@ opaServer:
...
+  # Set DefaultDecision: {} for deny by default.
```
Comment (shlomitk1) [inline charts/fybrik/values.yaml:446]: defaultDecision: {} ...

Comment (shlomitk1) [reply charts/fybrik/values.yaml:446]: Wouldn't defaultPolicy be a better name?

File: site/docs/tasks/using-opa.md [hunk 6]
```diff
@@ -44,16 +44,12 @@ An example of an object that may be returned by the rule above:
 
 ## Fybrik Default Policies
 
-Fybrik ***denys by default*** any request if no rule is triggered. This behavior can be changed to ***allow by default*** by creating the following rule and upload it to OPA using methods described in this page:
+Fybrik ***allows by default*** any request if no rule is triggered. This behavior can be changed to ***deny by default*** by altering altering the value of `opaServer.defaultDecision` to be `{}` during Fybrik's installation (the catalog is ignored for simplification):
```
Comment (shlomitk1) [inline site/docs/tasks/using-opa.md:47]: altering altering -> altering

**Comment by revit13** [comment]
@tomersolomon1 please also update the policy in https://fybrik.io/v1.2/samples/notebook-write/

File: charts/fybrik/templates/opa-server/opa-deployment.yaml [hunk 7]
```diff
@@ -72,6 +72,13 @@ spec:
             EOF
             {{ end }}
             {{ end }}
+            {{- if .Values.opaServer.defaultDecision}}
+            {{- range $policyName, $policy := .Values.opaServer.defaultDecision }} 
+            {{- printf "\ncat > /bootstrap/%s.rego <<EOF" $policyName  | indent 12 }}
+            {{ $policy | indent 12 }}
+            EOF
+            {{ end }}
```
Comment (revit13) [inline charts/fybrik/templates/opa-server/opa-deployment.yaml:80]: Shouldn't `charts/fybrik/files/opa-server/policy-lib/default_policy.rego` be updated?

Comment (shlomitk1) [reply charts/fybrik/templates/opa-server/opa-deployment.yaml:80]: @revit13 No, but maybe the name is misleading. This file is used for an internal implementation of opa-connector, maybe something like lib.rego or internal.rego

Comment (revit13) [reply charts/fybrik/templates/opa-server/opa-deployment.yaml:80]: @shlomitk1 thanks just updated my comment. I think that adding an explanation for the verdict will be helpful.

File: charts/fybrik/values.yaml [hunk 8]
```diff
@@ -442,6 +442,13 @@ opaServer:
...
+  defaultDecision:
```
Comment (roytman) [inline charts/fybrik/values.yaml:447]: I see a couple of issues here:

1. we increase the values file, theoretically we should provide an explanation to the entries
2. the default behavior should be a boolean: `true` - Allow, `false` - Deny. 
The rego transformation can be done behind of scene in `_helpers.tpl`

@shlomitk1 , can we remove the `bootstrapPolicies` ? slightly simplify the values file.

Comment (shlomitk1) [reply charts/fybrik/values.yaml:447]: Yes, we can.

File: charts/fybrik/files/opa-server/policy-lib/allowByDefault.rego [hunk 9]
```diff
@@ -0,0 +1 @@
+rule[{}] { true }
```
Comment (shlomitk1) [inline charts/fybrik/files/opa-server/policy-lib/allowByDefault.rego:1]: Add a new line at the end of the file

Comment (shlomitk1) [reply charts/fybrik/files/opa-server/policy-lib/allowByDefault.rego:1]: Also, the package name is missing.
```
 package dataapi.authz
 ```

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -122,6 +122,11 @@ jobs:
       env:
         K8S_VERSION: ${{ env.NEW_K8S_VERSION }}
       run: make run-integration-tests
+     - name: Test deny by defult
```
Comment (shlomitk1) [inline .github/workflows/build.yml:125]: default

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -122,6 +122,11 @@ jobs:
...
+      if: ${{ matrix.k8s_version == 'new' }}
+      env:
+        K8S_VERSION: ${{ env.NEW_K8S_VERSION }}
+      run: make run-deny-by-defult-test
```
Comment (shlomitk1) [inline .github/workflows/build.yml:129]: default

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 12]
```diff
@@ -226,3 +226,9 @@ func TestS3NotebookReadFlow(t *testing.T) {
 	}
 	record.Release()
 }
+
+// Test deny by default policy
+func TestS3ReadFlowDeny(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	g.Expect(2).To(gomega.Equal(3))
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:233]: ??

**Review by shlomitk1 — COMMENTED** [review]
Why do we need another CI test? There are too many of them already. Add the setup for deny to one of the existing checks and add the test to the existing test file.

**Comment by revit13** [comment]
@tomersolomon1 please update `Scenario one: write is forbidden due to governance restrictions` in https://fybrik.io/v1.2/samples/notebook-write/

File: Makefile [hunk 13]
```diff
@@ -139,7 +139,8 @@ test: pre-test
 
 .PHONY: run-integration-tests
 run-integration-tests: export VALUES_FILE=charts/fybrik/integration-tests.values.yaml
-run-integration-tests: export HELM_SETTINGS=--set "manager.solver.enabled=true"
+run-integration-tests: export HELM_SETTINGS=`--set "manager.solver.enabled=true" --set "coordinator.catalog=katalog" --set "opaServer.allowByDefault=false"`
```
Comment (shlomitk1) [inline Makefile:142]: integration tests work with mockup services, use one of the notebook tests instead.

File: Makefile [hunk 14]
```diff
@@ -149,6 +150,8 @@ run-integration-tests:
 	$(MAKE) -C pkg/helm test
 	$(MAKE) -C samples/rest-server test
 	$(MAKE) -C manager run-integration-tests
+	$(MAKE) setup-cluster
+	$(MAKE) -C manager run-deny-by-default-test
```
Comment (shlomitk1) [inline Makefile:154]: Integration tests do not use opa, thus any tests of opa policies have no effect.

File: Makefile [hunk 15]
```diff
@@ -157,12 +157,14 @@ run-notebook-readflow-tests:
...
-run-notebook-readflow-tests-katalog: export HELM_SETTINGS=--set "coordinator.catalog=katalog"
+run-notebook-readflow-tests-katalog: export HELM_SETTINGS=--set "coordinator.catalog=katalog" --set "opaServer.allowByDefault=false"
 run-notebook-readflow-tests-katalog: export VALUES_FILE=charts/fybrik/notebook-test-readflow.values.yaml
 run-notebook-readflow-tests-katalog: export CATALOGED_ASSET=fybrik-notebook-sample/data-csv
 run-notebook-readflow-tests-katalog: export DEPLOY_OPENMETADATA_SERVER=0
 run-notebook-readflow-tests-katalog: export USE_OPENMETADATA_CATALOG=0
 run-notebook-readflow-tests-katalog:
+	$(MAKE) setup-cluster
+	$(MAKE) -C manager run-deny-by-default-test
 	$(MAKE) setup-cluster
```
Comment (shlomitk1) [inline Makefile:168]: Why do you need additional cluster setup? It should be the same one.

File: manager/Makefile [hunk 16]
```diff
@@ -50,12 +50,16 @@ prep-test: wait_for_manager
...
 
-
 .PHONY: run-notebook-readflow-tests
 run-notebook-readflow-tests: prep-test
 	cd testdata/notebook/read-flow && ./setup.sh
 	NO_SIMULATED_PROGRESS=true USE_EXISTING_CONTROLLER=true USE_EXISTING_CLUSTER=true go test ./... $(TEST_OPTIONS) -run TestS3NotebookReadFlow -count 1
 
+.PHONY: run-deny-by-default-test
+run-deny-by-default-test: prep-test
+	cd testdata/notebook/read-flow && ./setup.sh
```
Comment (shlomitk1) [inline manager/Makefile:60]: I think we can skip the setup script and create the asset directly in the test code. There are some port-forwards that 1) are not required for the deny test and 2) I am not sure they can be done one after the other.
Another related comment: try to simplify the test. It is a deny, the plotter will never be created, no module will access the data, so creating the s3 service, uploading a file, etc. is an overkill.

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 17]
```diff
@@ -226,3 +226,145 @@ func TestS3NotebookReadFlow(t *testing.T) {
 	}
 	record.Release()
 }
+
+// Test deny by default policy
+func TestS3ReadFlowDeny(t *testing.T) {
+	t.Log("Started TestS3ReadFlowDeny")
+	if s, ok := os.LookupEnv("VALUES_FILE"); !ok ||
+		(s != "charts/fybrik/notebook-test-readflow.values.yaml" &&
+			s != "charts/fybrik/notebook-test-readflow.tls.values.yaml" &&
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:235]: Only charts/fybrik/notebook-test-readflow.values.yaml is set.

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 18]
```diff
@@ -226,3 +226,145 @@ func TestS3NotebookReadFlow(t *testing.T) {
...
+			s != "charts/fybrik/notebook-test-readflow.tls-system-cacerts.yaml") {
+		t.Skip("Only executed for notebook tests")
+	}
+	catalogedAsset, ok := os.LookupEnv("CATALOGED_ASSET")
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:239]: Why is it needed? You can take the asset id directly from spec.Data[0].DatasetID

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 19]
```diff
@@ -226,3 +226,145 @@ func TestS3NotebookReadFlow(t *testing.T) {
...
+	g.Eventually(func() error {
+		return k8sClient.Get(context.Background(), applicationKey, application)
+	}, timeout, interval).Should(gomega.Succeed())
+	fmt.Println("Expecting plotter to be constructed")
+	g.Eventually(func() *fapp.ResourceReference {
+		_ = k8sClient.Get(context.Background(), applicationKey, application)
+		return application.Status.Generated
+	}, timeout, interval).ShouldNot(gomega.BeNil())
+
+	// The plotter has to be created
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:335]: If you expect to get Deny, how will the plotter be created?

**Review by shlomitk1 — CHANGES_REQUESTED** [review]
I don't understand the newly added test. It is supposed to test deny by default, yet it uses bootstrap policies instead of no policies where the default could be triggered.
Also, this PR should be marked with ⚠️ because it breaks the backwards compatibility.

File: Makefile [hunk 20]
```diff
@@ -91,6 +91,7 @@ docker-mirror-read:
 undeploy-fybrik:
 	$(TOOLBIN)/helm uninstall fybrik-crd --namespace $(KUBE_NAMESPACE)
 	$(TOOLBIN)/helm uninstall fybrik --namespace $(KUBE_NAMESPACE)
+	$(TOOLBIN)/kubectl delete namespace $(KUBE_NAMESPACE) || true
```
Comment (shlomitk1) [inline Makefile:94]: Do not delete fybrik-system. This namespace is not controlled by fybrik/fybrik-crd releases. For example, vault is deployed in fybrik-system and should not be re-deployed each time.

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 21]
```diff
@@ -114,18 +115,95 @@ func TestS3NotebookReadFlow(t *testing.T) {
...
+		err = k8sClient.Get(context.Background(), applicationKey, application)
+		if err != nil {
+			return false
+		}
+		return application.Status.Ready
+	}, timeout, interval).Should(gomega.Equal(true))
+
+	modulesNamespace := plotter.Spec.ModulesNamespace
+	fmt.Printf("data access module namespace notebook test: %s\n", modulesNamespace)
+	g.Expect(application.Status.AssetStates[catalogedAsset].Conditions[DenyConditionIndex].Status).ToNot(gomega.Equal(v1.ConditionTrue))
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:167]: Better to check that in ReadyConditionIndex the status equals ConditionTrue.

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 22]
```diff
@@ -114,18 +115,95 @@ func TestS3NotebookReadFlow(t *testing.T) {
...
+	// Create reduct PII policy
+	g.Expect(readObjectFromFile("../../testdata/notebook/read-flow/pii-policy-cm.yaml", piiReadConfigMap)).ToNot(gomega.HaveOccurred())
+	piiReadConfigMapKey := client.ObjectKeyFromObject(piiReadConfigMap)
+	g.Expect(k8sClient.Create(context.Background(), piiReadConfigMap)).Should(gomega.Succeed())
+
+	fmt.Println("Expecting configmap to be created")
+	g.Eventually(func() error {
+		return k8sClient.Get(context.Background(), piiReadConfigMapKey, piiReadConfigMap)
+	}, timeout, interval).Should(gomega.Succeed())
+	fmt.Println("Expecting configmap to be constructed")
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:188]: Expecting policies to be compiled

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 23]
```diff
@@ -161,10 +239,12 @@ func TestS3NotebookReadFlow(t *testing.T) {
 		return application.Status.Ready
 	}, timeout, interval).Should(gomega.Equal(true))
 
-	modulesNamespace := plotter.Spec.ModulesNamespace
+	modulesNamespace = plotter.Spec.ModulesNamespace
 	fmt.Printf("data access module namespace notebook test: %s\n", modulesNamespace)
 
 	g.Expect(application.Status.AssetStates[catalogedAsset].Endpoint.Name).ToNot(gomega.BeEmpty())
+	g.Expect(application.Status.AssetStates[catalogedAsset].Conditions[DenyConditionIndex].Status).ToNot(gomega.Equal(v1.ConditionTrue))
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:246]: Same comment - check that it's ready

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 24]
```diff
@@ -41,10 +42,10 @@ type ArrowRequest struct {
 }
 
 func TestS3NotebookReadFlow(t *testing.T) {
-	if s, ok := os.LookupEnv("VALUES_FILE"); !ok ||
-		(s != "charts/fybrik/notebook-test-readflow.values.yaml" &&
-			s != "charts/fybrik/notebook-test-readflow.tls.values.yaml" &&
-			s != "charts/fybrik/notebook-test-readflow.tls-system-cacerts.yaml") {
+	valuesYaml, ok := os.LookupEnv("VALUES_FILE")
+	if !ok || (valuesYaml != "charts/fybrik/notebook-test-readflow.values.yaml" &&
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:46]: Create constants for the file names

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 25]
```diff
@@ -117,15 +118,108 @@ func TestS3NotebookReadFlow(t *testing.T) {
...
+		// Module installed by setup script directly from remote arrow-flight-module repository
+		// Installing application
+
+		g.Expect(readObjectFromFile("../../testdata/notebook/read-flow/fybrikapplication.yaml", application)).
+			ToNot(gomega.HaveOccurred())
+		application.ObjectMeta.Name += "-1"
+		application.Spec.Data[0].DataSetID = catalogedAsset
+		applicationKey = client.ObjectKeyFromObject(application)
+
+		// Create FybrikApplication and FybrikModule
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:140]: The comment is wrong - you do not create FybrikModule here

File: manager/controllers/app/fybrik_notebook_read_flow_test.go [hunk 26]
```diff
@@ -117,15 +118,108 @@ func TestS3NotebookReadFlow(t *testing.T) {
...
+			if err != nil {
+				return false
+			}
+			return application.Status.Ready
+		}, timeout, interval).Should(gomega.Equal(true))
+
+		modulesNamespace = plotter.Spec.ModulesNamespace
+		fmt.Printf("data access module namespace notebook test: %s\n", modulesNamespace)
+		g.Expect(application.Status.AssetStates[catalogedAsset].Conditions[ReadyConditionIndex].Status).To(gomega.Equal(v1.ConditionTrue))
+		g.Expect(application.Status.AssetStates[catalogedAsset].Endpoint.Name).ToNot(gomega.BeEmpty())
```
Comment (shlomitk1) [inline manager/controllers/app/fybrik_notebook_read_flow_test.go:175]: indentation?

Comment (tomersolomon1) [reply manager/controllers/app/fybrik_notebook_read_flow_test.go:175]: Which line is problematic?

Comment (shlomitk1) [reply manager/controllers/app/fybrik_notebook_read_flow_test.go:179]: ok, it appears as a very long line: 178, also - 189

File: site/docs/samples/notebook-write.md [hunk 27]
```diff
@@ -116,14 +116,21 @@ Below is the policy (written in [Rego](https://www.openpolicyagent.org/docs/late
...
-  input.action.destination != "neverland"
-  input.resource.metadata.columns[i].tags["PersonalData.Personal"]
-}
+rule[{"action": {"name":"Deny"}, "policy": description}] {
+    description := "Forbid writing sensitive data to theshire object-stores in datasets tagged with `finance`"
+    input.action.actionType == "write"
+    input.resource.metadata.tags.finance
+    input.action.destination == "theshire"
+    input.resource.metadata.columns[i].tags.sensitive
+    }
```
Comment (shlomitk1) [inline site/docs/samples/notebook-write.md:125]: Please leave the original tags. Others won't work with OpenMetadata catalog.

**Review by shlomitk1 — DISMISSED** [review]

**Review by shlomitk1 — APPROVED** [review]

**Pull request merged**